### PASSO A PASSO PARA CONFIGURAR O MYSQL PARA ACESSO REMOTO.

[RETIRADO DE:] https://www.codegrepper.com/code-examples/sql/mysql+grant+all+privileges+to+root

[RETIRADO DE:] https://dev.mysql.com/doc/refman/8.0/en/grant.html

[RETIRADO DE:] https://manage.accuwebhosting.com/knowledgebase/2200/How-to-Allow-Remote-Connection-to-MySQL-Server-through-Windows-Command-Prompt.html

[RETIRADO DE:] https://stackoverflow.com/questions/41645309/mysql-error-access-denied-for-user-rootlocalhost

Microsoft Windows [versão 10.0.19041.208]
(c) 2019 Microsoft Corporation. Todos os direitos reservados.

C:\Users\84538341>cd\

C:\>cd "Program Files\MySQL\MySQL Server 8.0\bin"


C:\Program Files\MySQL\MySQL Server 8.0\bin>mysql -u root -p
Enter password: **************
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 43
Server version: 8.0.28 MySQL Community Server - GPL

Copyright (c) 2000, 2022, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.


mysql> CREATE USER 'conectado'@'%' IDENTIFIED BY 'Wikazako!-1975';
Query OK, 0 rows affected (0.09 sec)

mysql> GRANT ALL PRIVILEGES ON *.* TO 'conectado'@'%' WITH GRANT OPTION;
Query OK, 0 rows affected (0.02 sec)

mysql> FLUSH PRIVILEGES;
Query OK, 0 rows affected (0.02 sec)

mysql> GRANT ALL PRIVILEGES ON world.* TO 'conectado'@'%';
Query OK, 0 rows affected (0.01 sec)

mysql> FLUSH PRIVILEGES;
Query OK, 0 rows affected (0.01 sec)

mysql> \h

For information about MySQL products and services, visit:
   http://www.mysql.com/
For developer information, including the MySQL Reference Manual, visit:
   http://dev.mysql.com/
To buy MySQL Enterprise support, training, or other products, visit:
   https://shop.mysql.com/

List of all MySQL commands:
Note that all text commands must be first on line and end with ';'
?         (\?) Synonym for `help'.
clear     (\c) Clear the current input statement.
connect   (\r) Reconnect to the server. Optional arguments are db and host.
delimiter (\d) Set statement delimiter.
ego       (\G) Send command to mysql server, display result vertically.
exit      (\q) Exit mysql. Same as quit.
go        (\g) Send command to mysql server.
help      (\h) Display this help.
notee     (\t) Don't write into outfile.
print     (\p) Print current command.
prompt    (\R) Change your mysql prompt.
quit      (\q) Quit mysql.
rehash    (\#) Rebuild completion hash.
source    (\.) Execute an SQL script file. Takes a file name as an argument.
status    (\s) Get status information from the server.
system    (\!) Execute a system shell command.
tee       (\T) Set outfile [to_outfile]. Append everything into given outfile.
use       (\u) Use another database. Takes database name as argument.
charset   (\C) Switch to another charset. Might be needed for processing binlog with multi-byte charsets.
warnings  (\W) Show warnings after every statement.
nowarning (\w) Don't show warnings after every statement.
resetconnection(\x) Clean session context.
query_attributes Sets string parameters (name1 value1 name2 value2 ...) for the next query to pick up.

For server side help, type 'help contents'

mysql> \q

In [1]:
import pymysql

import pandas as pd

In [2]:
conn=pymysql.connect(host='127.0.0.1',port=int(3306),user='conectado',passwd='Wikazako!-1975',db='classicmodels')

df=pd.read_sql_query("SELECT * FROM customers ",conn)

df.tail()

customerNumber                    customerName contactLastName  \
117             486    Motor Mint Distributors Inc.         Salazar   
118             487        Signal Collectibles Ltd.          Taylor   
119             489  Double Decker Gift Stores, Ltd           Smith   
120             495            Diecast Collectables          Franco   
121             496               Kelly's Gift Shop         Snowden   

    contactFirstName           phone        addressLine1 addressLine2  \
117             Rosa      2155559857   11328 Douglas Av.         None   
118              Sue      4155554312   2793 Furth Circle         None   
119          Thomas   (171) 555-7555     120 Hanover Sq.         None   
120          Valarie      6175552555      6251 Ingle Ln.         None   
121             Tony   +64 9 5555500  Arenales 1938 3'A'         None   

             city state postalCode      country  salesRepEmployeeNumber  \
117  Philadelphia    PA      71270          USA                  1323.0   
118      Brisbane    CA      94217          USA                  1165.0   
119        London  None    WA1 1DP           UK                  1501.0   
120        Boston    MA      51003          USA                  1188.0   
121    Auckland    None       None  New Zealand                  1612.0   

     creditLimit  
117      72600.0  
118      60300.0  
119      43300.0  
120      85100.0  
121     110000.0

In [3]:
df=pd.read_sql_query("SELECT user,authentication_string,plugin,host FROM mysql.user;",conn)

df

user                              authentication_string  \
0         conectado  $A$005$F#0*kDZCuQ?e,TCpzMRkV2iU74UJCRaSCt...   
1           luizcca  $A$005$j`7#&F-trsK2[JZicN3stx3BQV/THZTXEPi...   
2  mysql.infoschema  $A$005$THISISACOMBINATIONOFINVALIDSALTANDPASSW...   
3     mysql.session  $A$005$THISISACOMBINATIONOFINVALIDSALTANDPASSW...   
4         mysql.sys  $A$005$THISISACOMBINATIONOFINVALIDSALTANDPASSW...   
5              root  $A$005$QwS.>%1
'GG\nm).GQrBYlILDmyUHL9ZkbiDr...   

                  plugin       host  
0  caching_sha2_password          %  
1  caching_sha2_password          %  
2  caching_sha2_password  localhost  
3  caching_sha2_password  localhost  
4  caching_sha2_password  localhost  
5  caching_sha2_password  localhost

In [4]:
df=pd.read_sql_query("SELECT * FROM payments WHERE amount between 300 and 6000;",conn)

df.head()

customerNumber checkNumber paymentDate   amount
0             103    OM314933  2004-12-18  1676.14
1             121    FD317790  2003-10-28  1491.38
2             145    CN328545  2004-07-03  4710.73
3             148    DD635282  2004-08-11  2611.84
4             148    ME497970  2005-03-27  3516.04

In [5]:
df=pd.read_sql_query("select * from products \
where buyPrice > 32 and productLine <> 'Classic Cars' \
order by productVendor asc;",conn)

df.head()

productCode               productName       productLine productScale  \
0   S700_1138     The Schooner Bluenose             Ships        1:700   
1    S50_4713        2002 Yamaha YZR M1       Motorcycles         1:50   
2    S24_2841    1900s Vintage Bi-Plane            Planes         1:24   
3    S24_1578         1997 BMW R 1100 S       Motorcycles         1:24   
4    S24_2300  1962 Volkswagen Microbus  Trucks and Buses         1:24   

           productVendor                                 productDescription  \
0  Autoart Studio Design  All wood with canvas sails. Measures 31 1/2 in...   
1  Autoart Studio Design  Features rotating wheels , working kick stand....   
2  Autoart Studio Design  Hand crafted diecast-like metal bi-plane is re...   
3  Autoart Studio Design  Detailed scale replica with working suspension...   
4  Autoart Studio Design  This 1:18 scale die cast replica of the 1962 M...   

   quantityInStock  buyPrice    MSRP  
0             1897     34.00   66.67  
1              600     34.17   81.36  
2             5942     34.25   68.51  
3             7003     60.86  112.70  
4             2327     61.34  127.79

In [6]:
df=pd.read_sql_query("select * from customers;",conn)

df.head()

customerNumber                customerName contactLastName  \
0             103           Atelier graphique         Schmitt   
1             112          Signal Gift Stores            King   
2             114  Australian Collectors, Co.        Ferguson   
3             119           La Rochelle Gifts         Labrune   
4             121          Baane Mini Imports      Bergulfsen   

  contactFirstName         phone                  addressLine1 addressLine2  \
0          Carine     40.32.2555                54, rue Royale         None   
1             Jean    7025551838               8489 Strong St.         None   
2            Peter  03 9520 4555             636 St Kilda Road      Level 3   
3          Janine     40.67.8555  67, rue des Cinquante Otages         None   
4           Jonas     07-98 9555        Erling Skakkes gate 78         None   

        city     state postalCode    country  salesRepEmployeeNumber  \
0     Nantes      None      44000     France                  1370.0   
1  Las Vegas        NV      83030        USA                  1166.0   
2  Melbourne  Victoria       3004  Australia                  1611.0   
3     Nantes      None      44000     France                  1370.0   
4    Stavern      None       4110     Norway                  1504.0   

   creditLimit  
0      21000.0  
1      71800.0  
2     117300.0  
3     118200.0  
4      81700.0

In [7]:
df=pd.read_sql_query("select count(*) from customers;",conn)

df.head()

count(*)
0       122

[Accessing Databases with SQL Magic](https://notebooks.githubusercontent.com/view/ipynb?color_mode=auto&commit=dfc6722b3f78d5999758c1d2331bd18baa9d1e41&enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f676973742f7474616465737573692f36393232343230336330316666313037663733356436363439366266323661322f7261772f646663363732326233663738643539393937353863316432333331626431386261613964316534312f4d7953514c5f44617461626173652d53514c6d616769632e6970796e62&logged_in=false&nwo=ttadesusi%2F69224203c01ff107f735d66496bf26a2&path=MySQL_Database-SQLmagic.ipynb&repository_id=106162181&repository_type=Gist)

[SQL Interface within JupyterLab](https://www.datacamp.com/community/tutorials/sql-interface-within-jupyterlab?utm_source=adwords_ppc&utm_medium=cpc&utm_campaignid=1455363063&utm_adgroupid=65083631748&utm_device=c&utm_keyword=&utm_matchtype=&utm_network=g&utm_adpostion=&utm_creative=278443377095&utm_targetid=aud-299261629574:dsa-429603003980&utm_loc_interest_ms=&utm_loc_physical_ms=1001541&gclid=EAIaIQobChMIyYHK0dmO9wIVEEKRCh380gNEEAAYASAAEgJejfD_BwE)

[Treina ipython-sql](https://github.com/catherinedevlin)

[Connect to MySQL db from Jupyter notebook](https://stackoverflow.com/questions/50973191/connect-to-mysql-db-from-jupyter-notebook)

[SQL Certification Training Course](https://www.simplilearn.com/sql-database-training-course?utm_campaign=SQL&utm_medium=Description&utm_source=youtube)

In [8]:
df=pd.read_sql_query("select * from employees;",conn)

df.head()

employeeNumber   lastName firstName extension  \
0            1002     Murphy     Diane     x5800   
1            1056  Patterson      Mary     x4611   
2            1076   Firrelli      Jeff     x9273   
3            1088  Patterson   William     x4871   
4            1102     Bondur    Gerard     x5408   

                             email officeCode  reportsTo              jobTitle  
0     dmurphy@classicmodelcars.com          1        NaN             President  
1   mpatterso@classicmodelcars.com          1     1002.0              VP Sales  
2   jfirrelli@classicmodelcars.com          1     1002.0          VP Marketing  
3  wpatterson@classicmodelcars.com          6     1056.0  Sales Manager (APAC)  
4     gbondur@classicmodelcars.com          4     1056.0   Sale Manager (EMEA)



use sql_intro;





select emp_name, gender, dept, salary
from employees where salary >
(select salary from employees where emp_name = 'John');










# Stored Procedure

use sql_iq;
select * from players;

delimiter &&
create procedure top_players()
begin
select name, country, goals
from players where goals>6;
End &&
delimiter ;

call top_players();

#SP using IN

delimiter //
create procedure sp_sortBySalary(IN var int)
begin
select name, age, salary from emp_details
order by salary desc limit var;
end //
delimiter ;

call sp_sortBySalary(3);

delimiter //
create procedure update_salary(IN temp_name varchar(20), in
new_salary float)
begin
update emp_details set
salary = new_salary where name=temp_name;
end; //

select * from emp_details;

call update_salary('Marry', 80000);

#SP using OUT

delimiter //
create procedure sp_CountEmployees(OUT Total_Emps int)
begin
select count(name) into Total_Emps from emp_details
where sex='F';
end //
delimiter ;

call sp_CountEmployees(@F_emp);
select @F_emp as female_emps;

# Trigger in SQL

create table student
(st_roll int, age int, name varchar(30), mark float);

delimiter //
create trigger marks_verify_st
before insert on student
for each row
if new.mark < 0 then set new.mark=50;
end if; //

insert into student
values(501,10,'Ruth', 75.0),
(502,12, 'Mike', -20.5),
(503, 13, 'Dave', 90.0),
(504, 10, "Jacobs", -12.5);

select * from student;

# drop trigger marks_verify_st;

# Views in SQL

use classicmodels; 
select * from customers;

create view cust_details
as
select customerName, phone, city
from customers;

select * from cust_details;

select * from productlines;

create view product_description
as
select productName, quantityinstock,msrp,textdescription
from products as p inner join productlines as pl
on p.productline = pl.productline;

select * from product_description;

#Rename description
rename table product_description 
to vehicle_description; 

#Display views
show full tables
where table_type = 'VIEW';

#Delete view 

drop view cust_details;

#Windows fucntion

use sql_intro;
select * from employees;

select emp_name,age,dept,
sum(salary) over (partition by dept) as total_salary
from employees;

# Row number

select row_number() over(order by salary) as row_num,
emp_name, salary from employees order by salary;

create table demo (st_id int, st_name varchar(20));

insert into demo
values(101,'Shane'),
(102,'Bradley'),
(103,'Herath'),
(103,'Herath'),
(104, 'Nathan'),
(105,'Kevin'),
(105,'Kevin');

select * from demo;

select st_id, st_name, row_number() over 
(partition by st_id, st_name order by st_id) as row_num
from demo;

# Rank function

create table demo1(var_a int);

insert into demo1
value(101),(102),(103),(103),(104),(105),(106),(106),(107);

select var_a,
rank() over (order by var_a) as test_rank
from demo1;

#First value()

select emp_name, age, salary, first_value(emp_name)
over (order by salary desc) as highest_salary from employees;

select emp_name, dept, salary, first_value(emp_name)
over (partition by dept order by salary desc) as highest_salary
from employees;

# Subqueries

In [9]:
df=pd.read_sql_query("select * from employees;",conn)

df.head()

employeeNumber   lastName firstName extension  \
0            1002     Murphy     Diane     x5800   
1            1056  Patterson      Mary     x4611   
2            1076   Firrelli      Jeff     x9273   
3            1088  Patterson   William     x4871   
4            1102     Bondur    Gerard     x5408   

                             email officeCode  reportsTo              jobTitle  
0     dmurphy@classicmodelcars.com          1        NaN             President  
1   mpatterso@classicmodelcars.com          1     1002.0              VP Sales  
2   jfirrelli@classicmodelcars.com          1     1002.0          VP Marketing  
3  wpatterson@classicmodelcars.com          6     1056.0  Sales Manager (APAC)  
4     gbondur@classicmodelcars.com          4     1056.0   Sale Manager (EMEA)

In [ ]:
df=pd.read_sql_query("select firstName, lastName, email, salary \
from employees where salary > (select avg(salary) from employees);",conn)

df.head()

In [11]:
df=pd.read_sql_query("select * from products;",conn)

df.head()

productCode                            productName   productLine  \
0    S10_1678  1969 Harley Davidson Ultimate Chopper   Motorcycles   
1    S10_1949               1952 Alpine Renault 1300  Classic Cars   
2    S10_2016                  1996 Moto Guzzi 1100i   Motorcycles   
3    S10_4698   2003 Harley-Davidson Eagle Drag Bike   Motorcycles   
4    S10_4757                    1972 Alfa Romeo GTA  Classic Cars   

  productScale             productVendor  \
0         1:10           Min Lin Diecast   
1         1:10   Classic Metal Creations   
2         1:10  Highway 66 Mini Classics   
3         1:10         Red Start Diecast   
4         1:10   Motor City Art Classics   

                                  productDescription  quantityInStock  \
0  This replica features working kickstand, front...             7933   
1  Turnable front wheels; steering function; deta...             7305   
2  Official Moto Guzzi logos and insignias, saddl...             6625   
3  Model features, official Harley Davidson logos...             5582   
4  Features include: Turnable front wheels; steer...             3252   

   buyPrice    MSRP  
0     48.81   95.70  
1     98.58  214.30  
2     68.99  118.94  
3     91.02  193.66  
4     85.68  136.00

In [12]:
df=pd.read_sql_query("select * from products WHERE quantityInStock > 7000;",conn)

df.head()

productCode                            productName   productLine  \
0    S10_1678  1969 Harley Davidson Ultimate Chopper   Motorcycles   
1    S10_1949               1952 Alpine Renault 1300  Classic Cars   
2    S12_2823                       2002 Suzuki XREO   Motorcycles   
3    S12_3380                     1968 Dodge Charger  Classic Cars   
4    S12_4675                     1969 Dodge Charger  Classic Cars   

  productScale              productVendor  \
0         1:10            Min Lin Diecast   
1         1:10    Classic Metal Creations   
2         1:12       Unimax Art Galleries   
3         1:12  Welly Diecast Productions   
4         1:12  Welly Diecast Productions   

                                  productDescription  quantityInStock  \
0  This replica features working kickstand, front...             7933   
1  Turnable front wheels; steering function; deta...             7305   
2  Official logos and insignias, saddle bags loca...             9997   
3  1:12 scale model of a 1968 Dodge Charger. Hood...             9123   
4  Detailed model of the 1969 Dodge Charger. This...             7323   

   buyPrice    MSRP  
0     48.81   95.70  
1     98.58  214.30  
2     66.27  150.62  
3     75.16  117.44  
4     58.73  115.16

In [13]:
df=pd.read_sql_query("select * from orderdetails;",conn)

df.head()

orderNumber productCode  quantityOrdered  priceEach  orderLineNumber
0        10100    S18_1749               30     136.00                3
1        10100    S18_2248               50      55.09                2
2        10100    S18_4409               22      75.46                4
3        10100    S24_3969               49      35.29                1
4        10101    S18_2325               25     108.06                4

In [14]:
df=pd.read_sql_query("select productcode as 'Código do Produto', productname as Produto, msrp from products \
where productcode in (select productcode from orderdetails \
where priceeach < 100);",conn)

df.head()

Código do Produto                                Produto    msrp
0          S10_1678  1969 Harley Davidson Ultimate Chopper   95.70
1          S10_2016                  1996 Moto Guzzi 1100i  118.94
2          S12_3380                     1968 Dodge Charger  117.44
3          S12_3990                1970 Plymouth Hemi Cuda   79.80
4          S12_4473                      1957 Chevy Pickup  118.50

In [15]:
!jupyter nbconvert --to HTML conecta_query_MYSQL.ipynb

[NbConvertApp] Converting notebook conecta_query_MYSQL.ipynb to HTML
[NbConvertApp] Writing 616238 bytes to conecta_query_MYSQL.html
